osquery/sql/sqlite_util.cpp

```diff
@@ -459,41 +467,114 @@ Status QueryPlanner::applyTypes(TableColumns& columns) {
-        // Found a column name collision in the result.
-        VLOG(1) << "Detected overloaded column name " << column[i]
-                << " in query result consider using aliases";
+int readRows(sqlite3_stmt* prepared_statement, QueryDataTyped& results) {
+  int rc = SQLITE_ROW;
+  // First collect the column names and types
+  int num_columns = sqlite3_column_count(prepared_statement);
+  std::vector<std::string> colNames;
+  colNames.reserve(num_columns);
+  for (int i = 0; i < num_columns; i++) {
+    colNames.push_back(sqlite3_column_name(prepared_statement, i));
+  }
+
+  do {
+    RowTyped row;
+    for (int i = 0; i < num_columns; i++) {
+      switch (sqlite3_column_type(prepared_statement, i)) {
+      case SQLITE_INTEGER:
+        row[colNames[i]] =
+            static_cast<int64_t>(sqlite3_column_int64(prepared_statement, i));
+        break;
+      case SQLITE_FLOAT:
+        row[colNames[i]] = sqlite3_column_double(prepared_statement, i);
+        break;
+      case SQLITE_NULL:
+        row[colNames[i]] = FLAGS_nullvalue;
+        break;
+      default:
+        // Everything else (SQLITE_TEXT, SQLITE3_TEXT, SQLITE_BLOB) is
+        // obtained/conveyed as text/string
+        row[colNames[i]] = std::string(reinterpret_cast<const char*>(
+            sqlite3_column_text(prepared_statement, i)));
       }
-      r[column[i]] = (argv[i] != nullptr) ? argv[i] : FLAGS_nullvalue;
     }
-  }
-  (*qData).push_back(std::move(r));
-  return 0;
+    results.push_back(std::move(row));
+    rc = sqlite3_step(prepared_statement);
+  } while (SQLITE_ROW == rc);
+  return rc;
 }
 
-Status queryInternal(const std::string& q,
-                     QueryData& results,
+Status queryInternal(const std::string& query,
+                     QueryDataTyped& results,
                      const SQLiteDBInstanceRef& instance) {
-  char* err = nullptr;
-  auto lock = instance->attachLock();
-  sqlite3_exec(instance->db(), q.c_str(), queryDataCallback, &results, &err);
+  sqlite3_stmt* prepared_statement{nullptr}; /* Statement to execute. */
+
+  int rc = SQLITE_OK; /* Return Code */
+  const char* leftover_sql = nullptr; /* Tail of unprocessed SQL */
+  const char* sql = query.c_str(); /* SQL to be processed */
+
+  /* The big while loop.  One iteration per statement */
+  while ((sql[0] != 0) && (SQLITE_OK == rc)) {
+    const auto lock = instance->attachLock();
+
+    rc = sqlite3_prepare_v2(
+        instance->db(), sql, -1, &prepared_statement, &leftover_sql);
+    if (rc != SQLITE_OK) {
+      if (prepared_statement != nullptr) {
+        sqlite3_finalize(prepared_statement);
+      }
+      return Status::failure(sqlite3_errmsg(instance->db()));
+    } else {
+      if (prepared_statement == nullptr) {
+        /* this happens for a comment or white-space */
+        sql = leftover_sql;
+        while (isspace(sql[0])) {
+          sql++;
+        }
+        continue;
+      }
+
+      rc = sqlite3_step(prepared_statement);
```

Comment:
All this code from the first `sqlite3_step()` to `sqlite3_finalize()` should be in `readRows()` IMO. `readRows()` can then return a status that will be handle here.